apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientFactory.cs

```diff
@@ -0,0 +1,97 @@
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Auth;
+using Grpc.Core;
+
+namespace Google.Cloud.Spanner.Data
+{
+    internal interface ISpannerClientFactory
+    {
+        Task<SpannerClient> CreateClientAsync(ServiceEndpoint endpoint, ITokenAccess credential);
+    }
+
+    /// <summary>
+    /// For mockability and testability of the <see cref="ClientPool"/> class.
+    /// </summary>
+    internal class SpannerClientFactory : ISpannerClientFactory
+    {
+        internal static readonly ISpannerClientFactory Default = new SpannerClientFactory();
+
+        private SpannerClientFactory()
+        {
+        }
+
+        private static async Task<ChannelCredentials> CreateDefaultChannelCredentialsAsync()
+        {
+            var appDefaultCredentials = await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
+            if (appDefaultCredentials.IsCreateScopedRequired)
+            {
+                appDefaultCredentials = appDefaultCredentials.CreateScoped(SpannerClient.DefaultScopes);
+            }
+            return appDefaultCredentials.ToChannelCredentials();
+        }
+
+        /// <inheritdoc />
+        public async Task<SpannerClient> CreateClientAsync(ServiceEndpoint endpoint, ITokenAccess credential)
+        {
+            ChannelCredentials channelCredentials;
+
+            if (credential == null)
+            {
+                channelCredentials = await CreateDefaultChannelCredentialsAsync().ConfigureAwait(false);
+            }
+            else
+            {
+                channelCredentials = credential.ToChannelCredentials();
+            }
+
+            var channel = new Channel(
+                endpoint.Host,
+                endpoint.Port,
+                channelCredentials);
+            Logger.LogPerformanceCounterFn("SpannerClient.RawCreateCount", x => x + 1);
+
+            //Pull the timeout from spanner options.
+            //The option must be set before OpenAsync is called.
+            var callSettings = CallSettings.FromCallTiming(
+                CallTiming.FromRetry(
+                    new RetrySettings(
+                        SpannerSettings.GetDefaultRetryBackoff(),
+                        SpannerSettings.GetDefaultTimeoutBackoff(),
+                        Expiration.FromTimeout(SpannerOptions.Instance.Timeout),
+                        SpannerSettings.NonIdempotentRetryFilter
+                    )));
+
+            return SpannerClient.Create(
```

Comment:
actually these should all be *idempotent*. (I mistakenly thought all were marked nonidempotent in SpannerClient, but two *are* marked idempotent -Get/Delete Session).

However, Commit is the only API really at risk here which has a set of arguments which are not idempotent -- you can do a single commit call without a begintransaction to commit a set of mutations in one roundtrip.  This call, since it does not have a transactionid, is not idempotent.  However, I specifically avoid this because its not idempotent and instead use CommitSettings that always include a transactionid.  This is idempotent.

So actually I will purposely deviate from the base setting for a few of these APIs.
I'm not sure why CreateSession, ExecuteSql, Read, BeginTx are marked as nonidempotent.  Maybe vkit is using some general rule based on the operation type?